File: time_diff/tdiff.py
#!/usr/bin/env python
import fileinput
import re
import sys

FORTY = 40 * 60
TWENTY_FOUR = 24 * 60

def minutes_since_midnight(h,m,ap):
    return (h % 12 + (12 if ap == 'p' else 0))*60 + m

def report(m, type):
    if m <= 0: return
    hh, mm = m // 60, m % 60
    print(f"{hh}h {mm}m {type}")

running_total = 0
timeline = re.compile( 
    r'''^ \s*
     (?P<h1>\d{1,2}):(?P<m1>\d\d)(?P<ap1>[ap])m?
    \s+  - \s+
    (?P<h2>\d{1,2}):(?P<m2>\d\d)(?P<ap2>[ap])m?
    \s+
    (?P<lunch>\d{1,4})
    ''',re.X)

lines = fileinput.input()
for line in lines:
    line = line.strip()
    if len(line) == 0 or line.startswith('#'): continue
    parts = timeline.match(line)
    if not parts:
        print(f'!! Bad line {lines.lineno()}: <{line}>',file=sys.stderr)
        continue
    minutes = (minutes_since_midnight(int(parts['h2']),int(parts['m2']),parts['ap2']) - 
               minutes_since_midnight(int(parts['h1']),int(parts['m1']),parts['ap1']))
    if minutes < 0: minutes += TWENTY_FOUR
    minutes -= int(parts['lunch'])
    if minutes <= 0:
        print('!! No time on:', lines.lineno(), line, file=sys.stderr)
    
    remaining_reg = max(FORTY - running_total, 0)
    print(f"\nLine {lines.lineno()}: ",end=''); report(minutes,f"(from input {line})")
    report(min(minutes, remaining_reg), 'REG')
    report(max(minutes - remaining_reg, 0), 'OT')
    running_total += minutes

# now give grand totals
print("\n=========================================\nTOTALS:")
report(min(running_total, FORTY), "REG")
report(max(running_total - FORTY, 0), "OT")

File: time_diff/tdiff.rb
#-=======================================================
# tdiff.rb : add up times on a time sheet
#
# Example input:
# 
#  # Week of 5-01
#  8:30am - 5:00p 30   Monday
#  8:30am - 6:00p 30   Tuesday
#  8:30am - 5:00p 30   Wednesday
#  8:30am - 5:00p 30   Thursday
#  8:30am - 5:00p 30   Friday
#-=======================================================
FORTY = 40 * 60
TWENTY_FOUR = 24 * 60

def minutes_since_midnight(h,m,ap)
  (h % 12 + (ap == 'p' ? 12 : 0))*60 + m
end

def report(m, type)
  return if m <= 0
  hh, mm = (m/60).to_i, m % 60
  puts "#{hh}h #{mm}m #{type}"
end

running_total = 0
timeline = %r!
  ^ \s*
  (?<h1>\d{1,2}):(?<m1>\d\d)(?<ap1>[ap])m?
  \s+  - \s+
  (?<h2>\d{1,2}):(?<m2>\d\d)(?<ap2>[ap])m?
  \s+
  (?<lunch>\d{1,4})
!x

comments = %r!#.*$!
ARGF.each_line do |line|
  line.sub!(comments, '')
  line.strip!
  next if line.empty?

  parts = timeline.match(line)
  if not parts then
    STDERR.puts "!! Bad line #{ARGF.lineno}: <#{line}>"
    next
  end

  minutes = (minutes_since_midnight(parts['h2'].to_i,parts['m2'].to_i,parts['ap2']) - 
             minutes_since_midnight(parts['h1'].to_i,parts['m1'].to_i,parts['ap1']))
  minutes += TWENTY_FOUR if minutes < 0
  minutes -= parts['lunch'].to_i
  STDERR.puts "!! No time on: #{ARGF.lineno}: <#{line}>" if minutes <= 0

  remaining_reg = [FORTY - running_total, 0].max
  print "\nLine #{ARGF.lineno}: "; report(minutes, "(from input #{line})")
  report([minutes, remaining_reg].min, 'REG')
  report([minutes - remaining_reg, 0].max, 'OT')
  running_total += minutes
end

# now give grand totals
print <<~TOTS

===================================================
TOTALS:
TOTS

report([FORTY,running_total].min, "REG")
report([0, running_total - FORTY].max, "OT")
